### DISCUSSION TITLE
Service Discovery Endpoint Not Resolving from Client Side Code

### DISCUSSION THREAD

**Discussion by NapalmCodes** [microsoft/aspire#3818] [Q&A] [answered]
Probably expected, but while attempting to write a demo application for Aspire that utilizes SignalR, I have discovered using the service discovery endpoint in a web application (client side code) for an api service hosting the hub that it is unable to connect to the hub. The address `http://apiservice/myHub` I am going to guess is only resolvable within the orchestrator as the localhost address to the api service works fine. This is all well and good until we have multiple replicas of the api service running locally and want to "load balance" across the replicas.  I implemented the extension method from @davidfowl found [here](https://github.com/dotnet/aspire/issues/1356) for a different project in my solution (sends the message to the hub) with success but obviously that is in orchestrator and does not help with the JS client library.

I get CORS errors utilizing the `http://apiservice/myHub` address in chrome so I enabled those features on the API for allowing all origins to no success. Again, I realize this is probably a "bad error" if the intent is not to use the service discovery URLs locally. However, I do get a response from this URL locally which surprised me a bit (albeit not very useful):

`<html><head><title></title><meta name="revised" content="1.1.7" /></head><body></body></html>`

I could probably use the C# client on server and have a pass through hub to the client, but this feels a bit heavy handed. Another choice maybe is yarp or something else to reverse proxy traffic to all the replicas.

I am just curious what the recommendation from the team might be on this.

Here is my client side code currently (nothing fancy, straight from MS Learn):
```
"use strict";

let connection = new signalR.HubConnectionBuilder().withUrl("http://apiservice/myHub", { withCredentials: false }).build();

connection.on("ReceiveMessage", function (msg) {
    console.log(JSON.stringify(msg));
});

connection.start().then(function () {
    console.log("Connection started");
}).catch(function (err) {
    return console.error(err.toString());
});
```

Server side (apiservice):
```
using System.Text;
using System.Text.Json;
using Microsoft.AspNetCore.Mvc;

var builder = WebApplication.CreateBuilder(args);

// Add service defaults & Aspire components.
builder.AddServiceDefaults();

builder.Services.AddSignalR(o =>
{
    o.EnableDetailedErrors = true;
});

// Add services to the container.
builder.Services.AddProblemDetails();

builder.Services.AddCors(options =>
{
    options.AddDefaultPolicy(
        cpBuilder =>
        {
            cpBuilder.AllowAnyOrigin()
                .AllowAnyHeader()
                .AllowAnyMethod();
        });
});

var app = builder.Build();

app.UseCors();

// Configure the HTTP request pipeline.
app.UseExceptionHandler();

app.MapDefaultEndpoints();
app.MapHub<NotificationHub>("/myHub")
    .RequireCors();

app.Run();
```

**Comment by davidfowl** [comment]
Yes, this is expected. Service discovery only works when HTTP client is resolving the address. 

Do you have 2 different processes? 

![image](https://github.com/dotnet/aspire/assets/95136/e62a83a3-ee52-4541-b819-a8f22fa89e35)

**Reply by NapalmCodes** [reply]
Well my thought process was if I used the `.WithReplicas(2)` for instance on the "SignalR server" there will be no way to tell the JS client about all instances unless I give it all the URLs. It sounds like I will want the reverse proxy in front of this to direct the traffic, much like a Kubernetes "service" or "ingress" can do.

**Reply by davidfowl** [reply]
I'm just trying to understand the architecture. It's not clear to me what your application's topology is.

**Reply by svause-stryker** [reply]
Does this help?

![image](https://github.com/dotnet/aspire/assets/150449605/3219e8c7-fadc-4da9-8fbe-4eced6d99048)

API hosts the SignalR hub. Web is subscribed to the hub with the client JS library. Queue Consumer uses the C# client to send the message to subscribers.

**Reply by davidfowl** [reply]
Where's the browser and what does it connect to?

**Reply by NapalmCodes** [reply]
Browser connects to Web which is a razor pages app. It connects to API for REST calls and Hub communications.

**Reply by svause-stryker** [reply]
So I went down the path of trying to use Yarp as a proxy to forward traffic to the service discovery URL for API (http://apiservice). This isn't working as my guess is Yarp isn't service discovery enabled like `HttpClient` is. So I went down the rabbit hole of trying to use my Yarp proxy by reading environment variables for all the api service endpoints:

```
var keys = builder.Configuration.AsEnumerable()
    .Where(kvp => kvp.Key.StartsWith("services:apiservice:http:"))
    .Select(kvp => kvp.Key)
    .ToList();
```

That works with a single replica. However, I start an API Service with two replicas and neither instance spins up fully. My guess is this has something to do with the URL/port mapping being the same in the launchSettings.json launch profile (i.e.: applicationUrl).

```
var apiService = builder.AddProject<Projects.ApiService>("apiservice")
    .WithReference(cache)
    .WithReference(messaging)
    .WithReplicas(2);
```
 
Is there a way to launch two instances of the project with distinct ports?

I am not 100% sure I like this solution anyway as I am putting a proxy in front of another proxy (managed by Aspire) so clients can be round robin distributed to the api service replicas.

**Reply by davidfowl** [reply]
> Browser connects to Web which is a razor pages app. It connects to API for REST calls and Hub communications.

I would add the client to your diagram. Since its directly talking to the API but is rendered by the web project. That's why you need to enable cors.

**Reply by davidfowl** [reply]
A proxy is a good way to solve this, you can use YARP in your web project to forward the signalr calls to the API process.

**Reply by svause-stryker** [reply]
@davidfowl I put the proxy in place and it does work, but what am I missing in order to launch two local replicas of my API Service locally? Is there a way to spin up replicas with each having a unique port? The above code snippet using `.WithReplicas(x)` doesn't work for this and uses the port defined in the `launchSettings.json` for both. Maybe I am asking for too much out of the local orchestrator as it likely would be excessive to have two copies of the code running locally?

**Reply by davidfowl** [reply]
Using with replicas works because it will launch projects with a unique port.

**Reply by NapalmCodes** [reply]
It isn't though. The solution just hangs with neither api service becoming fully available and ultimately the whole thing crashes the apphost. Literally the only difference is the replicas part. I can post a repro if desired.

**Reply by davidfowl** [reply]
Sure post the repo!

**Reply by svause-stryker** [reply]
Before I post a reproduction I did a file new starter app from the template. Simply putting the line in for `.Replicas(2)` on the `apiservice` shows the same behavior:

![image](https://github.com/dotnet/aspire/assets/150449605/18d17a20-92d7-4406-884a-720bbb52c918)

Notice how it never starts the second instance and there are no logs for the first instance: 

![image](https://github.com/dotnet/aspire/assets/150449605/9a595d17-0ddd-4033-86fe-c349ab720332)

**Reply by davidfowl** [reply]
I got this:

<img width="1832" alt="image" src="https://github.com/dotnet/aspire/assets/95136/f40ab348-3a71-4999-9dc9-692e284477dd">

I'm using preview 6 though so it's possible something improved?

**Reply by svause-stryker** [reply]
Maybe? I am definitely on preview 5. I am guessing preview 6 isn't available for public installation yet? I am happy to try it when I can install it.

**Reply by davidfowl** [reply]
Wait a day 😄

**Reply by svause-stryker** [reply]
Got around to trying preview 6. No such luck same behavior. Is it possible the issue is MacOS specific? I am using Rider with the Aspire plugin:

![image](https://github.com/dotnet/aspire/assets/150449605/a12ae9d5-f4fa-415c-ab1f-121bc6ea00e6)

![image](https://github.com/dotnet/aspire/assets/150449605/7065037e-f8ee-4b62-bac1-2ada6b9aa0e9)

![image](https://github.com/dotnet/aspire/assets/150449605/0dc74bd0-3414-4017-b5f4-5be6d16f5410)

**Reply by svause-stryker** [reply, marked as answer]
Turns out using `dotnet run` works as expected. I am going to guess this is a rider plugin issue.

**Reply by davidfowl** [reply]
Make sure you file an issue
